src/AppBundle/Controller/DefaultController.php

```diff
@@ -53,21 +57,51 @@ public function registrationAction(Request $request)
     {
         // replace this example code with whatever you need
         if ($request->get('action') === 'register') {
-            $this->authenticationRegistrationService->register($request->get('NameID'));
-            return $this->authenticationRegistrationService->createRedirectResponse();
+            $this->registrationService->register($request->get('NameID'));
+            return $this->registrationService->replyToServiceProvider();
         }
 
         if ($request->get('action') === 'error') {
-            $this->authenticationRegistrationService->error($request->get('message'));
-            return $this->authenticationRegistrationService->createRedirectResponse();
+            $this->registrationService->reject($request->get('message'));
+            return $this->registrationService->replyToServiceProvider();
         }
 
-        $requiresRegistration = $this->authenticationRegistrationService->requiresRegistration();
+        $requiresRegistration = $this->registrationService->registrationRequired();
         $response = new Response(null, $requiresRegistration ? Response::HTTP_OK : Response::HTTP_BAD_REQUEST);
 
         return $this->render('AppBundle:default:registration.html.twig', [
             'requiresRegistration' => $requiresRegistration,
             'NameID' => uniqid('test-prefix-', 'test-entropy'),
         ], $response);
     }
+
+    /**
+     * @Route("/authentication", name="app_identity_authentication")
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function authenticationAction(Request $request)
+    {
+        $nameId = $this->authenticationService->getNameId();
+
+        // replace this example code with whatever you need
+        if ($request->get('action') === 'authenticate') {
+            // Implement the logic to verify authentication by the corresponding nameId.
+            $this->authenticationService->authenticate();
+            return $this->authenticationService->replyToServiceProvider();
+        }
+
+        if ($request->get('action') === 'error') {
```

Comment:
Just an idea: Maybe its a good thing to move the error action check to an earlier stage in this method? For example move it above the 'authenticate' check.